src/docs/binding-structs.md

```diff
@@ -0,0 +1,100 @@
+# Binding Structs
+
+Structs in Swift are value types that can have a number of forms each of which requires a certain amount of consideration: frozen/non-frozen and blitable and non-blitable.
+
+Structs that are frozen are guaranteed not to change. They have a fixed layout and are passed by value lowered up until the point when the size of the struct exceeds the number of registers dedicated for value types in Swift (currently 4 and unlikely to change).
+
+Non-frozen structs are passed by reference since the layout and size are not guaranteed.
+
+Similarly when calling members on frozen structs, the instance is passed by value as they can fit into registers. However for non-frozen structs the instance is passed by reference in the self pointer.
+
+Blittable structs are structs that can be correctly copied with `memcpy`. Essentially, if a struct contains a field which is heap allocated or contains a non-blittable data type, it will have issues with regards to reference counting. In addition, generic structs need to be considered non-blittable since they *may* contain non-blittable types, but we can't know that ahead of time.
+
+Structs in Swift are encouraged to be immutable. To help along with this, the compiler enforces a number of rules to prevent you from trying to mutate an immutable value. All arguments to a method are considered to be `let` bindings. The compiler won't let you modify public fields of struct that is defined with a `let` binding. Similarly, you cannot call a `mutating` method on a struct that is bound with `let`.
+
+Every value type in Swift comes with a compiler-generated Value Witness Table. Unlike other function tables in Swift (Protocol witness tables, vtables etc), the layout of value witness tables are well-defined. The value witness table is documented [here](binding-value-witness-table.md).
+
+## Language Parity Mismatches
+
+Structs in C# and Swift are very similar. They both represent value types with a concrete memory layout. The primary difference is that structs in Swift will execute code whenever it goes out of scope.
+
+Given the following Swift code:
+```swift
+public class Named {
+    private let name: String
+    public init(name n: String)
+    {
+        name = n
+    }
+    deinit
+    {
+        print("you killed \(name).")
+    }
+}
+
+public struct NamedHolder {
+    public init (name n: String) {
+        name = Named(name: n);
+    }
+    public var name: Named
+}
+
+public func runIt()
+{
+    var n1:NamedHolder = NamedHolder(name: "Corey")
+    let n2:NamedHolder = NamedHolder(name: "Ian")
+    n1 = n2
+}
+runIt()
+print("all done")
+```
+It will generate the following output:
+```
+you killed Corey.
+you killed Ian.
+all done
+```
+This shows how the destructor in the class gets executed when a struct gets overwritten ("you killed Corey") and when it goes out of scope. C# will do neither of this things.
+
+Swift allows 0-length structs, whereas C# does not.
```

Comment:
How are 0-length structs treated at the ABI layer? Are they expanded to 1-byte or are they omitted?

Reply:
0 length structs are omitted.